plugins/nf-google/src/main/nextflow/cloud/google/batch/GoogleBatchTaskHandler.groovy

```diff
@@ -586,17 +587,31 @@ class GoogleBatchTaskHandler extends TaskHandler implements FusionAwareTask {
      *
      * @return exit code if found, otherwise Integer.MAX_VALUE
      */
-    private Integer getExitCode(){
+    protected Integer getExitCode(){
         final events = client.getTaskStatus(jobId, taskId)?.getStatusEventsList()
         if( events ) {
-            final batchExitCode = events.stream().filter(ev -> ev.hasTaskExecution())
-                .max( (ev1, ev2) -> Long.compare(ev1.getEventTime().seconds, ev2.getEventTime().seconds) )
-                .map(ev -> ev.getTaskExecution().getExitCode())
-                .orElse(null)
-            if( batchExitCode != null )
-                return batchExitCode
-        }
-        // fallback to read
+            // Find the most recent event that contains a TaskExecution with an exit code.
+            // Events are not guaranteed to be in chronological order, so we iterate through
+            // all of them and track the one with the highest timestamp.
+            StatusEvent latestEvent = null
+            long latestTime = Long.MIN_VALUE
+            for( StatusEvent ev : events ) {
+                // Only consider events that have task execution info (which contains exit code)
+                if( ev.hasTaskExecution() ) {
+                    final long eventTime = ev.getEventTime().getSeconds()
+                    if( eventTime > latestTime ) {
+                        latestTime = eventTime
+                        latestEvent = ev
+                    }
+                }
+            }
+            // Return the exit code from the most recent task execution event
+            if( latestEvent != null ) {
+                return latestEvent.getTaskExecution().getExitCode()
```

Comment:
There is a corner case where if there is a task execution event but with exit code is null. In this case, it will not fallback to readExitFile. Not sure if it is a possible case, but better to protect.
```suggestion
            if( latestEvent?.getTaskExecution()?.getExitCode() != null ) {
                return latestEvent.getTaskExecution().getExitCode()
```